<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Tyler Grimes" />

<meta name="date" content="2021-12-03" />

<title>Package datasets</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Package datasets</h1>
<h4 class="author">Tyler Grimes</h4>
<h4 class="date">2021-12-03</h4>


<div id="TOC">
<ul>
<li><a href="#reactome-pathways"><span class="toc-section-number">1</span> Reactome pathways</a>
<ul>
<li><a href="#subsetting-pathways-using-pathway-names"><span class="toc-section-number">1.1</span> Subsetting pathways using pathway names</a></li>
<li><a href="#subsetting-pathways-using-genes"><span class="toc-section-number">1.2</span> Subsetting pathways using genes</a></li>
<li><a href="#p53_pathways-data"><span class="toc-section-number">1.3</span> <code>p53_pathways</code> data</a></li>
</ul></li>
<li><a href="#gene-expression-example"><span class="toc-section-number">2</span> Gene expression example</a>
<ul>
<li><a href="#meso-data"><span class="toc-section-number">2.1</span> <code>meso</code> data</a>
<ul>
<li><a href="#clinical-dataset"><span class="toc-section-number">2.1.1</span> Clinical dataset</a></li>
<li><a href="#rna-seq-dataset"><span class="toc-section-number">2.1.2</span> RNA-seq dataset</a></li>
<li><a href="#finalizing"><span class="toc-section-number">2.1.3</span> Finalizing</a></li>
</ul></li>
</ul></li>
<li><a href="#biomart"><span class="toc-section-number">3</span> biomaRt</a></li>
</ul>
</div>

<p>This vignette shows how the package datasets were created. The package data includes a list of cancer-related Reactome pathways, <code>p53_pathways</code>, and an example gene expression dataset, <code>glioma</code>. These are provided to be used in examples and for illustrative purposes.</p>
<div id="reactome-pathways" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Reactome pathways</h1>
<p>A list of Reaectome pathways is provided for Homo sapiens (human). This list was initialized using the <code>dnapath::get_reactome_pathways()</code> function.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>pathway_list <span class="ot">&lt;-</span> dnapath<span class="sc">::</span><span class="fu">get_reactome_pathways</span>(<span class="st">&quot;human&quot;</span>)</span></code></pre></div>
<pre><code>## Obtaining reactome pathway information for species: Homo sapiens 
## Combining pathways that have greater than 90 % overlap.</code></pre>
<p>This list contains the gene sets for 885 pathways. Here are the first two pathways:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>pathway_list[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span></code></pre></div>
<pre><code>## $`Purine salvage`
##  [1] &quot;100&quot;    &quot;132&quot;    &quot;161823&quot; &quot;1633&quot;   &quot;1716&quot;   &quot;270&quot;    &quot;271&quot;    &quot;272&quot;   
##  [9] &quot;2766&quot;   &quot;3251&quot;   &quot;353&quot;    &quot;4860&quot;   &quot;51292&quot; 
## 
## $`Nucleotide salvage`
##  [1] &quot;100&quot;    &quot;132&quot;    &quot;151531&quot; &quot;161823&quot; &quot;1633&quot;   &quot;1716&quot;   &quot;1890&quot;   &quot;270&quot;   
##  [9] &quot;271&quot;    &quot;272&quot;    &quot;2766&quot;   &quot;3251&quot;   &quot;353&quot;    &quot;4860&quot;   &quot;51292&quot;  &quot;54963&quot; 
## [17] &quot;7083&quot;   &quot;7084&quot;   &quot;7371&quot;   &quot;7378&quot;   &quot;8226&quot;   &quot;83549&quot;  &quot;978&quot;</code></pre>
<p>The list names correspond to the Reactome pathway name. Each pathway provides the entrezgene IDs of all genes in that pathway. For example, from the output above, we see that the first pathway in the list is the “purine salvage” pathway, and it contains the genes: 100, 132, 161823, etc.</p>
<p>In practice, all of these pathways can be used in the <code>dnapath()</code> analysis. However, in some studies we may only be interested in a subset of this list. For example, if we are studying a cancer-related gene expression dataset, we may have a specific interest in cancer-related pathways. There are two ways to obtain a subset from the pathway list: (1) using the pathway names or (2) using the pathway genes. These two approaches are demonstrated in turn.</p>
<div id="subsetting-pathways-using-pathway-names" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> Subsetting pathways using pathway names</h2>
<p>Suppose we are interested in pathways that are related to “p53” or “pi3k” regulatory processes (these are known cancer-related pathways). One option is to subset pathways based on whether or not the pathway names include these keywords. The <code>grep()</code> function is used to do this:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Obtain index of pathways containing &quot;p53&quot; or &quot;pi3k&quot; in their name.</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>index_pathways <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">&quot;((p53)|(pi3k))&quot;</span>, <span class="fu">names</span>(pathway_list),</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><code>grep</code> uses the regular expression “((p53)|(pi3k))”, which says to identify all pathway names that include the phrase “p53” or “pi3k”. However, we don’t want to leave out pathways that use capital “P53” or “PI3K”, so we set <code>ignore.case = TRUE</code> to ignore the case of the pathway names.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset onto the &quot;p53&quot; and &quot;pi3k&quot; pathways.</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>cancer_pathways <span class="ot">&lt;-</span> pathway_list[index_pathways]</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>cancer_pathways[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>] <span class="co"># Print out the first three</span></span></code></pre></div>
<pre><code>## $`CD28 dependent PI3K/Akt signaling`
##  [1] &quot;10000&quot;  &quot;117145&quot; &quot;1326&quot;   &quot;207&quot;    &quot;208&quot;    &quot;2475&quot;   &quot;253260&quot; &quot;2534&quot;  
##  [9] &quot;3932&quot;   &quot;5170&quot;   &quot;5290&quot;   &quot;5295&quot;   &quot;5296&quot;   &quot;55615&quot;  &quot;57761&quot;  &quot;64223&quot; 
## [17] &quot;79109&quot;  &quot;8503&quot;   &quot;9020&quot;   &quot;940&quot;    &quot;941&quot;    &quot;942&quot;   
## 
## $`G beta:gamma signalling through PI3Kgamma`
##  [1] &quot;10000&quot;  &quot;10681&quot;  &quot;146850&quot; &quot;207&quot;    &quot;208&quot;    &quot;23533&quot;  &quot;2782&quot;   &quot;2783&quot;  
##  [9] &quot;2784&quot;   &quot;2785&quot;   &quot;2786&quot;   &quot;2787&quot;   &quot;2788&quot;   &quot;2790&quot;   &quot;2791&quot;   &quot;2792&quot;  
## [17] &quot;2793&quot;   &quot;387&quot;    &quot;5170&quot;   &quot;51764&quot;  &quot;5294&quot;   &quot;54331&quot;  &quot;55970&quot;  &quot;59345&quot; 
## [25] &quot;94235&quot; 
## 
## $`Regulation of TP53 Degradation (See also: Regulation of TP53 Expression and Degradation)`
##  [1] &quot;10000&quot;  &quot;1017&quot;   &quot;1029&quot;   &quot;11200&quot;  &quot;117584&quot; &quot;1616&quot;   &quot;207&quot;    &quot;208&quot;   
##  [9] &quot;2475&quot;   &quot;253260&quot; &quot;4193&quot;   &quot;4194&quot;   &quot;472&quot;    &quot;51230&quot;  &quot;5170&quot;   &quot;5515&quot;  
## [17] &quot;5516&quot;   &quot;5518&quot;   &quot;5519&quot;   &quot;5527&quot;   &quot;55615&quot;  &quot;6233&quot;   &quot;64223&quot;  &quot;6446&quot;  
## [25] &quot;7157&quot;   &quot;7311&quot;   &quot;7314&quot;   &quot;7316&quot;   &quot;7874&quot;   &quot;79109&quot;  &quot;80196&quot;  &quot;890&quot;   
## [33] &quot;8900&quot;   &quot;900&quot;    &quot;9099&quot;   &quot;983&quot;    &quot;639&quot;</code></pre>
<p>Subsetting on these cancer-related pathways results in 19 pathways.</p>
<p>In general, the pathway list obtained from the <code>dnapath::get_reactome_pathways()</code> function can be subset based on the pathway names, as illustrated in this example. Here, we focused on cancer pathways, but the regular expression “((p53)|(pi3k))” can be replaced to search for any desired pathways. To learn more about the <code>grep()</code> function and regular expression, see <a href="https://bookdown.org/rdpeng/rprogdatascience/regular-expressions.html">chapter 17</a> of R Programming for Data Science by Roger Peng.</p>
</div>
<div id="subsetting-pathways-using-genes" class="section level2" number="1.2">
<h2><span class="header-section-number">1.2</span> Subsetting pathways using genes</h2>
<p>As in the previous example, suppose we are interested in pathways that are related to “p53” or “pi3k”. However, this time we want to identify <em>all</em> pathways that involve the genes, not just pathways that are named after them. (Note that some pathways may contain the p53 gene but not have “p53” in its name.)</p>
<p>Since the pathway list uses entrezgene IDs, the first step is to find the entrezgene ID for each gene of interest. The ID for P53 is <a href="https://www.ncbi.nlm.nih.gov/gene?cmd=Retrieve&amp;dopt=full_report&amp;list_uids=7157">7157</a>. PI3K actually refers to a family of enzymes; one class of genes in this family include PIK3CA, PIK3CB, PIK3CD, and PIK3CG, which have the entrezgene IDs <a href="https://www.ncbi.nlm.nih.gov/gene?cmd=Retrieve&amp;dopt=full_report&amp;list_uids=5290">5290</a>, <a href="https://www.ncbi.nlm.nih.gov/gene?cmd=Retrieve&amp;dopt=full_report&amp;list_uids=5291">5291</a>, <a href="https://www.ncbi.nlm.nih.gov/gene?cmd=Retrieve&amp;dopt=full_report&amp;list_uids=5293">5293</a>, and <a href="https://www.ncbi.nlm.nih.gov/gene?cmd=Retrieve&amp;dopt=full_report&amp;list_uids=5294">5294</a>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a vector of all entrezgene IDs that we are interested in.</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>genes_of_interest <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">7157</span>, <span class="dv">5290</span>, <span class="dv">5291</span>, <span class="dv">5293</span>, <span class="dv">5294</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Obtain index of pathways containing any genes of interest.</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>index_pathways <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">sapply</span>(pathway_list, <span class="cf">function</span>(pathway) {</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">any</span>(genes_of_interest <span class="sc">%in%</span> pathway)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>}))</span></code></pre></div>
<p>The <code>sapply()</code> function is used to loop over each pathway in the pathway list, and for each we determine if any of the genes of interest are in the pathway’s gene set. We end up with a set of indices identifying which pathways contain at least one of the five targeted genes.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset onto pathways containing p53 gene or pi3k family of genes.</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>cancer_pathways <span class="ot">&lt;-</span> pathway_list[index_pathways]</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>cancer_pathways[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>] <span class="co"># Print out the first three</span></span></code></pre></div>
<pre><code>## $`Activation of BH3-only proteins`
##  [1] &quot;10000&quot;  &quot;10018&quot;  &quot;10971&quot;  &quot;140735&quot; &quot;1869&quot;   &quot;207&quot;    &quot;208&quot;    &quot;23368&quot; 
##  [9] &quot;27113&quot;  &quot;2810&quot;   &quot;5366&quot;   &quot;5533&quot;   &quot;5534&quot;   &quot;5599&quot;   &quot;572&quot;    &quot;572&quot;   
## [17] &quot;596&quot;    &quot;637&quot;    &quot;7027&quot;   &quot;7029&quot;   &quot;7157&quot;   &quot;7159&quot;   &quot;7161&quot;   &quot;7529&quot;  
## [25] &quot;7531&quot;   &quot;7532&quot;   &quot;7533&quot;   &quot;7534&quot;   &quot;8626&quot;   &quot;8655&quot;   &quot;90427&quot; 
## 
## $`Signaling by ERBB2`
##  [1] &quot;10000&quot;  &quot;10193&quot;  &quot;10273&quot;  &quot;10718&quot;  &quot;10718&quot;  &quot;11140&quot;  &quot;145957&quot; &quot;145957&quot;
##  [9] &quot;1729&quot;   &quot;1839&quot;   &quot;1839&quot;   &quot;1950&quot;   &quot;1950&quot;   &quot;1956&quot;   &quot;1956&quot;   &quot;2064&quot;  
## [17] &quot;2064&quot;   &quot;2065&quot;   &quot;2065&quot;   &quot;2066&quot;   &quot;2066&quot;   &quot;2069&quot;   &quot;2069&quot;   &quot;207&quot;   
## [25] &quot;208&quot;    &quot;2534&quot;   &quot;2549&quot;   &quot;26469&quot;  &quot;2885&quot;   &quot;2885&quot;   &quot;2886&quot;   &quot;3084&quot;  
## [33] &quot;3084&quot;   &quot;3265&quot;   &quot;3265&quot;   &quot;3320&quot;   &quot;3845&quot;   &quot;3845&quot;   &quot;387&quot;    &quot;4145&quot;  
## [41] &quot;4893&quot;   &quot;4893&quot;   &quot;51072&quot;  &quot;5290&quot;   &quot;5295&quot;   &quot;5335&quot;   &quot;5578&quot;   &quot;5580&quot;  
## [49] &quot;5581&quot;   &quot;55914&quot;  &quot;5753&quot;   &quot;5782&quot;   &quot;5782&quot;   &quot;6233&quot;   &quot;6464&quot;   &quot;6464&quot;  
## [57] &quot;6654&quot;   &quot;6654&quot;   &quot;6714&quot;   &quot;685&quot;    &quot;685&quot;    &quot;7311&quot;   &quot;7314&quot;   &quot;7316&quot;  
## [65] &quot;7525&quot;   &quot;8065&quot;   &quot;9101&quot;   &quot;9542&quot;   &quot;9542&quot;  
## 
## $`CD28 co-stimulation`
##  [1] &quot;10000&quot;  &quot;117145&quot; &quot;1326&quot;   &quot;207&quot;    &quot;208&quot;    &quot;2475&quot;   &quot;253260&quot; &quot;2534&quot;  
##  [9] &quot;2885&quot;   &quot;3932&quot;   &quot;4067&quot;   &quot;5058&quot;   &quot;5062&quot;   &quot;5063&quot;   &quot;5170&quot;   &quot;5290&quot;  
## [17] &quot;5295&quot;   &quot;5296&quot;   &quot;55615&quot;  &quot;57761&quot;  &quot;5879&quot;   &quot;64223&quot;  &quot;6714&quot;   &quot;7409&quot;  
## [25] &quot;7525&quot;   &quot;79109&quot;  &quot;8503&quot;   &quot;9020&quot;   &quot;940&quot;    &quot;9402&quot;   &quot;941&quot;    &quot;942&quot;   
## [33] &quot;998&quot;</code></pre>
<p>Subsetting results in 80 pathways that contain at least one of the five cancer-related genes of interest. In general, any set of entrezgene IDs can be used to subset the pathways obtained from the <code>dnapath::get_reactome_pathways()</code> function, as illustrated in this example.</p>
</div>
<div id="p53_pathways-data" class="section level2" number="1.3">
<h2><span class="header-section-number">1.3</span> <code>p53_pathways</code> data</h2>
<p>The <code>dnapath</code> package provides an example pathway list named <code>p53_pathways</code>. This is a small list provided for illustrative purposes and for use in examples. It includes T53 Reactome pathways, which was created using the steps described previously.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Obtain index of pathways containing &quot;p53&quot; in their name.</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>index_pathways <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">&quot;(p53)&quot;</span>, <span class="fu">names</span>(pathway_list),<span class="at">ignore.case =</span> <span class="cn">TRUE</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>p53_pathways <span class="ot">&lt;-</span> pathway_list[index_pathways]</span></code></pre></div>
<p>The dataset was compressed and saved for the package using the code:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Not run: Only for building the package.</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>usethis<span class="sc">::</span><span class="fu">use_data</span>(p53_pathways)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>tools<span class="sc">::</span><span class="fu">resaveRdaFiles</span>(<span class="fu">file.path</span>(<span class="st">&quot;data&quot;</span>), <span class="at">compress =</span> <span class="st">&quot;auto&quot;</span>)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>tools<span class="sc">::</span><span class="fu">checkRdaFiles</span>(<span class="fu">file.path</span>(<span class="st">&quot;data&quot;</span>))</span></code></pre></div>
</div>
</div>
<div id="gene-expression-example" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Gene expression example</h1>
<p>A small gene expression dataset is provided in <code>dnapath</code>. This example data is a Mesothelioma dataset containing gene expression and clinical data that were generated by <a href="https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga">The Cancer Genome Atlas</a> (TCGA) and is downloaded using the <a href="http://www.linkedomics.org/data_download/TCGA-GBMLGG/">LinkedOmics portal</a>.</p>
<div id="meso-data" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> <code>meso</code> data</h2>
<p>The first step is to download and process the clinical and gene expression dataset from LinkedOmics.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>file_clinical <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;http://linkedomics.org/data_download/TCGA-MESO/Human__&quot;</span>,</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">&quot;TCGA_MESO__MS__Clinical__Clinical__01_28_2016__BI__&quot;</span>,</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">&quot;Clinical__Firehose.tsi&quot;</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>file_rnaseq <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;http://linkedomics.org/data_download/TCGA-MESO/Human__&quot;</span>,</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>                      <span class="st">&quot;TCGA_MESO__UNC__RNAseq__HiSeq_RNA__01_28_2016__BI__Gene&quot;</span>,</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>                      <span class="st">&quot;__Firehose_RSEM_log2.cct.gz&quot;</span>)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>clinical <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_tsv</span>(file_clinical)</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>rnaseq <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_tsv</span>(file_rnaseq)</span></code></pre></div>
<div id="clinical-dataset" class="section level3" number="2.1.1">
<h3><span class="header-section-number">2.1.1</span> Clinical dataset</h3>
<p>The differential network analysis performed using <code>dnapath()</code> compares the gene-gene association structure between two groups (populations). In the Mesothelioma dataset, the clinical data can be used to identify two subgroups of interest. In this example dataset, the tumor stage (classified as stage i, ii, iii, or iv) will be used to partition the data. In practice, <code>dnapath()</code> could be used make pair-wise comparisons across these tumor stages, or perhaps some stages can be combined together. But for the purpose of creating a small, example dataset, we will only look at tumor stages ii and iv.</p>
<p>The first step is to process the downloaded data so that it contains one column of sample IDs and one column of tumor stage.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># First, we can glimpse at the raw data to see how it is structured.</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>clinical[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code></pre></div>
<pre><code>## # A tibble: 5 × 3
##   attrib_name       TCGA.3H.AB3O TCGA.3H.AB3S
##   &lt;chr&gt;             &lt;chr&gt;        &lt;chr&gt;       
## 1 years_to_birth    58           69          
## 2 pathologic_stage  stageiii     stageii     
## 3 pathology_T_stage t3           t2          
## 4 pathology_N_stage n0           n0          
## 5 pathology_M_stage m0           m0</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The columns contain samples, and rows contain different variables</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co"># We want the transpose of this.</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>variable_names <span class="ot">&lt;-</span> clinical<span class="sc">$</span>attrib_name <span class="co"># Store the variable names.</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Transpose the matrix so that columns correspond to variables.</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>clinical <span class="ot">&lt;-</span> <span class="fu">t</span>(clinical[, <span class="sc">-</span><span class="dv">1</span>]) <span class="co"># Don&#39;t include the &quot;attrib_name&quot; column as a row.</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(clinical) <span class="ot">&lt;-</span> variable_names <span class="co"># Rename columns using variable names.</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Glimpse at the transposed dataset:</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>clinical[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]</span></code></pre></div>
<pre><code>##              years_to_birth pathologic_stage pathology_T_stage
## TCGA.3H.AB3O &quot;58&quot;           &quot;stageiii&quot;       &quot;t3&quot;             
## TCGA.3H.AB3S &quot;69&quot;           &quot;stageii&quot;        &quot;t2&quot;             
## TCGA.3U.A98E &quot;71&quot;           &quot;stageiv&quot;        &quot;t4&quot;             
## TCGA.3U.A98G &quot;71&quot;           &quot;stageiv&quot;        &quot;t4&quot;             
##              pathology_N_stage
## TCGA.3H.AB3O &quot;n0&quot;             
## TCGA.3H.AB3S &quot;n0&quot;             
## TCGA.3U.A98E &quot;n2&quot;             
## TCGA.3U.A98G &quot;n0&quot;</code></pre>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Now that the rows correspond to subjects and columns to variables,</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="co"># the next step is to add a new column that contains the sample IDs. </span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>clinical <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="at">ID =</span> <span class="fu">rownames</span>(clinical), clinical)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(clinical) <span class="ot">&lt;-</span> <span class="cn">NULL</span> <span class="co"># Remove the row names.</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Glimpse at the data.</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>clinical[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]</span></code></pre></div>
<pre><code>##      ID             years_to_birth pathologic_stage pathology_T_stage
## [1,] &quot;TCGA.3H.AB3O&quot; &quot;58&quot;           &quot;stageiii&quot;       &quot;t3&quot;             
## [2,] &quot;TCGA.3H.AB3S&quot; &quot;69&quot;           &quot;stageii&quot;        &quot;t2&quot;             
## [3,] &quot;TCGA.3U.A98E&quot; &quot;71&quot;           &quot;stageiv&quot;        &quot;t4&quot;             
## [4,] &quot;TCGA.3U.A98G&quot; &quot;71&quot;           &quot;stageiv&quot;        &quot;t4&quot;</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For the example dataset, we are only interested in ID and tumor stage.</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>clinical <span class="ot">&lt;-</span> clinical[, <span class="fu">c</span>(<span class="st">&quot;ID&quot;</span>, <span class="st">&quot;pathologic_stage&quot;</span>)]</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co"># The two groups that will be compared are stage 2 and stage 4.</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset rows onto those that have stage 2 or stage 4.</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>clinical <span class="ot">&lt;-</span> clinical[clinical[, <span class="st">&quot;pathologic_stage&quot;</span>] <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;stageii&quot;</span>, <span class="st">&quot;stageiv&quot;</span>), ]</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>clinical[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code></pre></div>
<pre><code>##      ID             pathologic_stage
## [1,] &quot;TCGA.3H.AB3S&quot; &quot;stageii&quot;       
## [2,] &quot;TCGA.3U.A98E&quot; &quot;stageiv&quot;       
## [3,] &quot;TCGA.3U.A98G&quot; &quot;stageiv&quot;       
## [4,] &quot;TCGA.SC.AA5Z&quot; &quot;stageiv&quot;       
## [5,] &quot;TCGA.UT.A88D&quot; &quot;stageii&quot;</code></pre>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We are left with only stage 2 and stage 4 samples:</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(clinical[, <span class="dv">2</span>])</span></code></pre></div>
<pre><code>## 
## stageii stageiv 
##      16      16</code></pre>
<p>The final dataset contains 16 samples for group 1 (tumor stage ii) and 16 samples for group 2 (tumor stage iv). Note that the groups don’t need equal sample sizes, but they happen to be equal in these data.</p>
</div>
<div id="rna-seq-dataset" class="section level3" number="2.1.2">
<h3><span class="header-section-number">2.1.2</span> RNA-seq dataset</h3>
<p>The second step is to process the gene expression dataset. The LinkedOmics portal provides normalized counts from the RNA-seq experiments (the data are not raw reads, which must be aligned and annotated in order to obtain gene expression counts). For the purpose of creating a small example dataset, these data will be subset onto the 32 samples obtained in the previous section, and onto the 160 genes in the <code>cancer_pathway</code> list.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># First, we can glimps at the raw data to see how it is structured.</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>rnaseq[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## # A tibble: 5 × 5
##   attrib_name TCGA.3H.AB3K TCGA.3H.AB3L TCGA.3H.AB3M TCGA.3H.AB3O
##   &lt;chr&gt;              &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;
## 1 A1BG               6.25          5.02        8.02          6.57
## 2 A1CF               0.687         0           0.462         0   
## 3 A2BP1              6.21          3.74        6.92          1.30
## 4 A2LD1              5.26          5.70        7.49          7.03
## 5 A2ML1              0.687         0           0             0</code></pre>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># As with the clinical data, we need to transpose the raw data so that</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co"># samples are in the rows.</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>gene_symbols <span class="ot">&lt;-</span> rnaseq<span class="sc">$</span>attrib_name <span class="co"># Store the gene names</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>rnaseq <span class="ot">&lt;-</span> <span class="fu">t</span>(rnaseq[, <span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(rnaseq) <span class="ot">&lt;-</span> gene_symbols</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>rnaseq[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##                A1BG   A1CF  A2BP1  A2LD1  A2ML1
## TCGA.3H.AB3K 6.2508 0.6869 6.2128 5.2615 0.6869
## TCGA.3H.AB3L 5.0232 0.0000 3.7424 5.6967 0.0000
## TCGA.3H.AB3M 8.0151 0.4622 6.9154 7.4918 0.0000
## TCGA.3H.AB3O 6.5719 0.0000 1.2995 7.0304 0.0000
## TCGA.3H.AB3S 5.2770 0.0000 6.7363 7.0906 0.0000</code></pre>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check if all 32 clinical samples contain gene expression data.</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(clinical[, <span class="st">&quot;ID&quot;</span>] <span class="sc">%in%</span> <span class="fu">rownames</span>(rnaseq))</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset the rnaseq data onto those in the clinical data</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>rnaseq <span class="ot">&lt;-</span> rnaseq[<span class="fu">rownames</span>(rnaseq) <span class="sc">%in%</span> clinical[, <span class="st">&quot;ID&quot;</span>], ]</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Finally, we must make sure the two dataset are aligned.</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a><span class="co"># There are many ways to do this, here is one:</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a><span class="co">#   For each ID in the clinical dataset, find the corresponding row in rnaseq.</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a><span class="co">#   The rows are then reordered to match the IDs in clinical.</span></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>rnaseq <span class="ot">&lt;-</span> rnaseq[<span class="fu">sapply</span>(clinical[, <span class="st">&quot;ID&quot;</span>], </span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>                        <span class="cf">function</span>(ID) <span class="fu">which</span>(<span class="fu">rownames</span>(rnaseq) <span class="sc">==</span> ID)), ]</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Check that the IDs match:</span></span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(<span class="fu">rownames</span>(rnaseq) <span class="sc">==</span> clinical[, <span class="st">&quot;ID&quot;</span>])</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>rnaseq[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##                A1BG   A1CF  A2BP1  A2LD1 A2ML1
## TCGA.3H.AB3S 5.2770 0.0000 6.7363 7.0906     0
## TCGA.3U.A98E 4.1398 0.0000 6.0186 5.8955     0
## TCGA.3U.A98G 7.9146 0.0000 2.7608 6.8913     0
## TCGA.SC.AA5Z 5.8400 0.0000 7.3606 5.7489     0
## TCGA.UT.A88D 6.1513 0.6561 4.2798 5.1120     0</code></pre>
<p>The final step of processing the gene expression data is to subset the columns to the genes contained in the <code>cancer_pathway</code> list. However, the raw data provide gene symbols rather than entrezgene IDs (which the Reactome pathways use). So, we must either relabel the gene symbols in the <code>rnaseq</code> dataset to entrezgene IDs, or relabel the entrezgene IDs in <code>cancer_pathway</code> to contain gene symbols. Either of these operations can be easily carried out using the <code>symbol_to_entrez()</code>, <code>entrez_to_symbol()</code>, and <code>rename_genes()</code> functions provided in the package. These two approaches are briefly demonstrated next.</p>
<blockquote>
<p><strong>Note:</strong> Internet connection is required to connect to biomaRt, which the <code>entrez_to_symbol</code> and <code>symbol_to_entrez</code> methods use to map entrezgene IDs and gene symbols. The <code>dir_save</code> argument can be set when running these to save the ID mapping obtained from biomaRt in the specified directory. This way, the mapping can be obtained once (using the internet connection) and accessed from memory in all future calls of these functions. The temporary directory <code>tempdir()</code> is used in this example.</p>
</blockquote>
<p>First, we show how to convert gene symbols in the gene expression dataset to entrezgene IDs.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert gene symbols -&gt; entrezgene IDs</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>gene_symbols <span class="ot">&lt;-</span> <span class="fu">colnames</span>(rnaseq) <span class="co"># Extract the gene symbols to relabel.</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>gene_mat <span class="ot">&lt;-</span> <span class="fu">symbol_to_entrez</span>(gene_symbols, <span class="st">&quot;human&quot;</span>,</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>                             <span class="at">dir_save =</span> <span class="fu">tempdir</span>()) <span class="co"># Obtain mapping.</span></span></code></pre></div>
<pre><code>##  - loading gene info from /var/folders/vb/115zgf4x28vf40dccy9mj5hw0000gn/T//RtmpKgzacH/entrez_to_hsapiens.rds</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>gene_mat[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code></pre></div>
<pre><code>##   hgnc_symbol entrezgene_id
## 1        A1BG             1
## 2        A1CF         29974
## 3       A2BP1            -1
## 4       A2LD1            -1
## 5       A2ML1        144568</code></pre>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The rename_genes() method is a multipurpose function that can be used to rename </span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="co"># the genes in a vector (the current case), a pathway list, or the &#39;dnapath_list&#39;</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="co"># object that is returned from dnapath() after performing the differential</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="co"># network analysis.</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>gene_entrez <span class="ot">&lt;-</span> <span class="fu">rename_genes</span>(gene_symbols, gene_mat) <span class="co"># Rename the symbols.</span></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(rnaseq) <span class="ot">&lt;-</span> gene_entrez <span class="co"># Update the columns using the entrezgene IDs.</span></span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>rnaseq[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>] <span class="co"># Columns now contain entrezgene IDs</span></span></code></pre></div>
<pre><code>##                   1  29974     -1     -1 144568
## TCGA.3H.AB3S 5.2770 0.0000 6.7363 7.0906      0
## TCGA.3U.A98E 4.1398 0.0000 6.0186 5.8955      0
## TCGA.3U.A98G 7.9146 0.0000 2.7608 6.8913      0
## TCGA.SC.AA5Z 5.8400 0.0000 7.3606 5.7489      0
## TCGA.UT.A88D 6.1513 0.6561 4.2798 5.1120      0</code></pre>
<p>This is the approach used to create the <code>meso</code> data that is provided in this package. In general, it is advised to use entrezgene IDs throughout the analysis pipeline, and <strong>only convert to symbols at the very end</strong>. Entrezgene IDs are unique identifiers for each gene, whereas gene symbols can sometimes be ambiguous. This ambiguity can lead to information loss if used early on the analysis. For example, notice the third column in above output, the “A2BP1” gene, was not mapped to an entrezgene ID. This is because “A2BP1” is an alias for the RBFOX1 gene. The <code>symbol_to_entrez()</code> function obtains a mapping for RBFOX1 but not for the alias A2BP1. Because an entrezgene ID was not identified for A2BP1, it will not be identified in any of the pathways during the differential network analysis.</p>
<p>Gene symbols should only be considerd at the end of the analysis. It can be conventient to rename the IDs into symbols, since symbols are more recognizable for viewing and interpreting the results.</p>
<p><strong>Important note:</strong> In the Mesothelioma dataset, we are utilizing summarized gene expression counts as opposed to raw RNA-sequencing reads. If the raw reads are unavailable, then the user may not be able to control whether or not the genes are entrezgene IDs or gene symbols. In this case, we must be aware of the limitations and take caution if many gene symbols fail to map to entrezgene IDs. However, if raw reads are available, then it is recommended to align and annotate the reads using entrezgene IDs when obtaining the expression counts.</p>
<p>To finish the processing, the columns of <code>rnaseq</code> are subset onto the genes in <code>p53_pathways</code>.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset the columns onto only those genes contained in &#39;p53_pathways&#39;</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>index_genes <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">colnames</span>(rnaseq) <span class="sc">%in%</span> <span class="fu">get_genes</span>(p53_pathways))</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>rnaseq <span class="ot">&lt;-</span> rnaseq[, index_genes]</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(rnaseq) <span class="co"># 32 samples with 150 genes.</span></span></code></pre></div>
<pre><code>## [1]  32 149</code></pre>
<p>For completeness, we will show how to convert the entrezgene IDs in the pathway list into gene symbols.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert entrezgene IDs -&gt; gene symbols</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>gene_entrez <span class="ot">&lt;-</span> <span class="fu">get_genes</span>(p53_pathways) <span class="co"># Extract genes from pathway list.</span></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>gene_mat <span class="ot">&lt;-</span> <span class="fu">entrez_to_symbol</span>(gene_entrez, <span class="st">&quot;human&quot;</span>, </span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>                             <span class="at">dir_save =</span> <span class="fu">tempdir</span>()) <span class="co"># Obtain mapping.</span></span></code></pre></div>
<pre><code>##  - loading gene info from /var/folders/vb/115zgf4x28vf40dccy9mj5hw0000gn/T//RtmpKgzacH/entrez_to_hsapiens.rds</code></pre>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>gene_mat[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code></pre></div>
<pre><code>##   entrezgene_id hgnc_symbol
## 1         10000        AKT3
## 2          1017        CDK2
## 3          1029      CDKN2A
## 4         11200       CHEK2
## 5        117584        RFFL</code></pre>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert the entrezgene IDs into gene symbols in the pathway list.</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>new_pathway_list <span class="ot">&lt;-</span> <span class="fu">rename_genes</span>(p53_pathways, gene_mat) </span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>new_pathway_list[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>] <span class="co"># Print the first two pathways.</span></span></code></pre></div>
<pre><code>## $`Regulation of TP53 Degradation (See also: Regulation of TP53 Expression and Degradation)`
##  [1] &quot;AKT3&quot;    &quot;CDK2&quot;    &quot;CDKN2A&quot;  &quot;CHEK2&quot;   &quot;RFFL&quot;    &quot;DAXX&quot;    &quot;AKT1&quot;   
##  [8] &quot;AKT2&quot;    &quot;MTOR&quot;    &quot;RICTOR&quot;  &quot;MDM2&quot;    &quot;MDM4&quot;    &quot;ATM&quot;     &quot;PHF20&quot;  
## [15] &quot;PDPK1&quot;   &quot;PPP2CA&quot;  &quot;PPP2CB&quot;  &quot;PPP2R1A&quot; &quot;PPP2R1B&quot; &quot;PPP2R5C&quot; &quot;PRR5&quot;   
## [22] &quot;RPS27A&quot;  &quot;MLST8&quot;   &quot;SGK1&quot;    &quot;TP53&quot;    &quot;UBA52&quot;   &quot;UBB&quot;     &quot;UBC&quot;    
## [29] &quot;USP7&quot;    &quot;MAPKAP1&quot; &quot;RNF34&quot;   &quot;CCNA2&quot;   &quot;CCNA1&quot;   &quot;CCNG1&quot;   &quot;USP2&quot;   
## [36] &quot;CDK1&quot;    &quot;PRDM1&quot;  
## 
## $`Regulation of TP53 Activity through Acetylation`
##  [1] &quot;AKT3&quot;    &quot;CHD3&quot;    &quot;CHD4&quot;    &quot;EP300&quot;   &quot;AKT1&quot;    &quot;AKT2&quot;    &quot;BRD1&quot;   
##  [8] &quot;BRPF3&quot;   &quot;BRD7&quot;    &quot;HDAC1&quot;   &quot;HDAC2&quot;   &quot;ING2&quot;    &quot;PIN1&quot;    &quot;PIP4K2A&quot;
## [15] &quot;MBD3&quot;    &quot;PML&quot;     &quot;GATAD2A&quot; &quot;MAP2K6&quot;  &quot;GATAD2B&quot; &quot;RBBP4&quot;   &quot;RBBP7&quot;  
## [22] &quot;MEAF6&quot;   &quot;TP53&quot;    &quot;BRPF1&quot;   &quot;PIP4K2C&quot; &quot;KAT6A&quot;   &quot;PIP4K2B&quot; &quot;ING5&quot;   
## [29] &quot;PIP4P1&quot;  &quot;MTA2&quot;</code></pre>
</div>
<div id="finalizing" class="section level3" number="2.1.3">
<h3><span class="header-section-number">2.1.3</span> Finalizing</h3>
<p>The <code>dnapath()</code> function requires either a list of two gene expression datasets (each corresponding to a different group), or a single gene expression data and a ‘group’ vector that indicates which rows belong to which group. The <code>meso</code> data is formatted using the second option. The current <code>rnaseq</code> dataset that we have processed contains samples from both groups; the group information can be obtained from the <code>clinical</code> dataset. These two pieces of information can be conveniently stored as a list.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>meso <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">gene_expression =</span> rnaseq,</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">groups =</span> clinical[, <span class="st">&quot;pathologic_stage&quot;</span>])</span></code></pre></div>
<p>The list was compressed and saved for the package using the code:</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Not run: Only for building the package.</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>usethis<span class="sc">::</span><span class="fu">use_data</span>(meso)</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>tools<span class="sc">::</span><span class="fu">resaveRdaFiles</span>(<span class="fu">file.path</span>(<span class="st">&quot;data&quot;</span>), <span class="at">compress =</span> <span class="st">&quot;auto&quot;</span>)</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>tools<span class="sc">::</span><span class="fu">checkRdaFiles</span>(<span class="fu">file.path</span>(<span class="st">&quot;data&quot;</span>))</span></code></pre></div>
</div>
</div>
</div>
<div id="biomart" class="section level1" number="3">
<h1><span class="header-section-number">3</span> biomaRt</h1>
<p>The <code>biomaRt</code> package is used to link gene symbols with entrezgene IDs. The <code>entrez_to_symbol()</code> and <code>symbol_to_entrez()</code> functions require an internet connection to connect with the biomaRt database.</p>
<p>This package contains an instance of the mapping between HGNC gene symbols and entregene IDs, which will be used if internet access is not available or if the <code>biomaRt</code> package is not installed. However, this dataset will not be applicable to species other than <em>hsapiens</em> and may not contain the most up-to-date mapping. A warning is given when it is used.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Not run: Internet access and `biomaRt` package is required.</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>mart <span class="ot">&lt;-</span> biomaRt<span class="sc">::</span><span class="fu">useMart</span>(<span class="at">biomart =</span> <span class="st">&quot;ensembl&quot;</span>, </span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>                         <span class="at">dataset =</span> <span class="st">&quot;hsapiens_gene_ensembl&quot;</span>)</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>biomart_hsapiens <span class="ot">&lt;-</span> biomaRt<span class="sc">::</span><span class="fu">getBM</span>(<span class="at">attributes =</span> <span class="fu">c</span>(<span class="st">&quot;entrezgene_id&quot;</span>, <span class="st">&quot;hgnc_symbol&quot;</span>),</span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">mart =</span> mart)</span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr) <span class="co"># For pipe operator &quot;%&gt;%&quot;.</span></span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a>biomart_hsapiens <span class="sc">%&gt;%</span></span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(entrezgene_id)) <span class="sc">%&gt;%</span></span>
<span id="cb50-11"><a href="#cb50-11" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(entrezgene_id) <span class="sc">%&gt;%</span></span>
<span id="cb50-12"><a href="#cb50-12" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">summarise</span>(<span class="fu">across</span>(<span class="fu">everything</span>(), dplyr<span class="sc">::</span>first)) <span class="ot">-&gt;</span></span>
<span id="cb50-13"><a href="#cb50-13" aria-hidden="true" tabindex="-1"></a>  biomart_hsapiens</span></code></pre></div>
<p>This mapping was compressed and saved for the package using the code:</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Not run: Only for building the package.</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>usethis<span class="sc">::</span><span class="fu">use_data</span>(biomart_hsapiens)</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>tools<span class="sc">::</span><span class="fu">resaveRdaFiles</span>(<span class="fu">file.path</span>(<span class="st">&quot;data&quot;</span>), <span class="at">compress =</span> <span class="st">&quot;auto&quot;</span>)</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>tools<span class="sc">::</span><span class="fu">checkRdaFiles</span>(<span class="fu">file.path</span>(<span class="st">&quot;data&quot;</span>))</span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
